server/embedded/src/org/labkey/embedded/LabKeyServer.java

```diff
@@ -0,0 +1,472 @@
+            }
+
+            private List<ContextResource> getDataSourceResources(ContextProperties props) throws ConfigException
+            {
+                List<ContextResource> dataSourceResources = new ArrayList<>();
+                var numOfDataResources = props.getUrl().size();
+
+                if (numOfDataResources != props.getDataSourceName().size() ||
+                        numOfDataResources != props.getDriverClassName().size() ||
+                        numOfDataResources != props.getUsername().size() ||
+                        numOfDataResources != props.getPassword().size())
+                {
+                    throw new ConfigException("DataSources not configured properly. Must have all the required properties for all datasources: dataSourceName, driverClassName, userName, password, url");
+                }
+
+                for (int i = 0; i < numOfDataResources; i++)
+                {
+                    ContextResource dataSourceResource = new ContextResource();
+                    dataSourceResource.setName(props.getDataSourceName().get(i));
+                    dataSourceResource.setAuth("Container");
+                    dataSourceResource.setType(DataSource.class.getName());
+                    dataSourceResource.setProperty("driverClassName", props.getDriverClassName().get(i));
+                    dataSourceResource.setProperty("url", props.getUrl().get(i));
+                    dataSourceResource.setProperty("password", props.getPassword().get(i));
+                    dataSourceResource.setProperty("username", props.getUsername().get(i));
+                    // TODO : 8021 move this properties to application.properties
+                    dataSourceResource.setProperty("maxTotal", "20");
+                    dataSourceResource.setProperty("maxIdle", "10");
+                    dataSourceResource.setProperty("maxWaitMillis", "120000");
+                    dataSourceResource.setProperty("accessToUnderlyingConnectionAllowed", "true");
+                    dataSourceResource.setProperty("validationQuery", "SELECT 1");
+                    dataSourceResources.add(dataSourceResource);
+                }
+
+                return  dataSourceResources;
+            }
+
+            private ContextResource getMailResource()
+            {
+                // Get session/mail properties
+                MailProperties mailProps = smtpSource();
+                ContextResource mailResource = new ContextResource();
+                mailResource.setName("mail/Session");
+                mailResource.setAuth("Container");
+                mailResource.setType("javax.mail.Session");
+                mailResource.setProperty("mail.smtp.host", mailProps.getSmtpHost());
+                mailResource.setProperty("mail.smtp.user", mailProps.getSmtpUser());
+                mailResource.setProperty("mail.smtp.port", mailProps.getSmtpPort());
+
+                if (mailProps.getSmtpFrom() != null &&
```

Comment:
I don't think these should be all-or-nothing. Setting a from address shouldn't mean that you have to have a passowrd or custom socket factory, or vice versa.

Reply:
ok, changed